File: Algorithm_62_Unique Paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0 for _ in range(m)] for _ in range(n)]
        dp[0][0] = 1
        for i in range(n):
            for j in range(m):
                if i == 0 and j == 0:
                    continue
                leftNum = 0 if i == 0 else dp[i - 1][j]
                topNum = 0 if j == 0 else dp[i][j - 1]
                dp[i][j] = leftNum + topNum
        return dp[n - 1][m - 1]

File: Algorithm_62_Unique Paths.java
// 思路：动态规划。dp[i][j]存储走到[i][j]格子所拥有的unique path总数
// dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
// 注意i == 0和j == 0的边界情况
// 空间复杂度可以从O(m * n)优化到O(m)

class Solution {
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];
        dp[0][0] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i > 0) {
                    dp[i][j] += dp[i - 1][j];
                }
                if (j > 0) {
                    dp[i][j] += dp[i][j - 1];
                }
            }
        }
        return dp[m - 1][n - 1];
    }
}